### PR TITLE
Allow `cuco::arrow_filter_policy` to accept a custom implementation of `xxhash_64`

### PR DESCRIPTION
**Pull request by mhaseeb123**
This PR allows  `cuco::arrow_filter_policy` to accept a custom implementation of `xxhash_64` hasher or use `cuco::xxhash_64` if nothing provided. Related to https://github.com/rapidsai/cudf/issues/17164 and https://github.com/rapidsai/cudf/pull/17393

### CODE REVIEW

File: include/cuco/detail/hash_functions/xxhash.cuh [hunk 1]
```diff
@@ -20,8 +20,22 @@
...
+// Helper trait to check if a type is `span` like.
+// i.e., if it has `data()` and `size()` functions
+template <typename T, typename = void>
+struct is_span_like : std::false_type {};
+
+// Specialization for `span` like type.
+template <typename T>
+struct is_span_like<
+  T,
+  std::void_t<decltype(std::declval<T>().data()), decltype(std::declval<T>().size())>>
```
Comment (mhaseeb123) [inline include/cuco/detail/hash_functions/xxhash.cuh]: Got this snippet from NVCode (4o in disguise)

File: include/cuco/detail/hash_functions/xxhash.cuh [hunk 2]
```diff
@@ -20,8 +20,22 @@
...
 #include <cuda/std/cstddef>
+#include <cuda/std/span>
 
 #include <cstdint>
+#include <type_traits>
+
+// Helper trait to check if a type is `span` like.
+// i.e., if it has `data()` and `size()` functions
+template <typename T, typename = void>
+struct is_span_like : std::false_type {};
```
Comment (PointKernel) [inline include/cuco/detail/hash_functions/xxhash.cuh]:
```suggestion
template <typename T, typename = void>
struct is_span_like : cuda::std::false_type {};
```

File: include/cuco/detail/hash_functions/xxhash.cuh [hunk 3]
```diff
@@ -20,8 +20,22 @@
...
+
+// Specialization for `span` like type.
+template <typename T>
+struct is_span_like<
+  T,
+  std::void_t<decltype(std::declval<T>().data()), decltype(std::declval<T>().size())>>
+  : std::true_type {};
```
Comment (PointKernel) [inline include/cuco/detail/hash_functions/xxhash.cuh]:
```suggestion
  cuda::std::void_t<decltype(cuda::std::declval<T>().data()), decltype(cuda::std::declval<T>().size())>>
  : cuda::std::true_type {};
```

File: include/cuco/detail/hash_functions/xxhash.cuh [hunk 4]
```diff
@@ -283,6 +297,18 @@ struct XXHash_64 {
     }
   }
 
+  /**
+   * @brief Returns a hash value for its `span` like argument, as a value of type `result_type`.
+   *
+   * @param key The input argument to hash
+   * @return The resulting hash value for `span` like `key`
+   */
+  template <typename T = Key, typename = std::enable_if_t<is_span_like<T>::value>>
```
Comment (PointKernel) [inline include/cuco/detail/hash_functions/xxhash.cuh]:
```suggestion
  template <typename T = Key, typename = cuda::std::enable_if_t<is_span_like<T>::value>>
```

**Review by PointKernel — COMMENTED** [review]
Looks great! Let's use `cuda::std::` utilities when possible for all host-device functions.

**Comment by mhaseeb123** [comment]
Converting to draft to add a couple of bitwise tests for `string` types with Arrow's implementation. Please feel free to look at the code already pushed here as any addition will only be tests.

File: include/cuco/bloom_filter_policies.cuh [hunk 5]
```diff
@@ -28,9 +29,12 @@ namespace cuco {
  * fingerprint.
  *
  * @tparam Key The type of the values to generate a fingerprint for.
+ * @tparam Hash Hash function used to generate a key's fingerprint. By default, cuco::xxhash_64 will
+ * be used.
+ *
  */
-template <class Key>
-using arrow_filter_policy = detail::arrow_filter_policy<Key>;
+template <class Key, class Hash = cuco::xxhash_64<Key>>
```
Comment (sleeepyjack) [inline include/cuco/bloom_filter_policies.cuh]: One potential problem with this is that a user might pass in a hash function that is incompatible with this policy (e.g. a 32 or 128bit hasher).

File: include/cuco/detail/hash_functions/xxhash.cuh [hunk 6]
```diff
@@ -20,8 +20,29 @@
...
 
 #include <cuda/std/cstddef>
+#include <cuda/std/span>
 
 #include <cstdint>
+#include <type_traits>
+
+// Helper trait to check if a type is `span` like.
+template <typename T, typename = void>
+struct is_span_like : cuda::std::false_type {};
```
Comment (sleeepyjack) [inline include/cuco/detail/hash_functions/xxhash.cuh]: Do we require something "span-like", i.e., a continuous memory segment or would it be sufficient if the type is random-access-iterable?

Comment (mhaseeb123) [reply include/cuco/detail/hash_functions/xxhash.cuh]: Discussed offline. @PointKernel added a specialized impl support `cudf::string_view` in https://github.com/rapidsai/cudf/pull/17393 fulfiling our needs.

File: include/cuco/bloom_filter_policies.cuh (hunk 5 above)
Comment (mhaseeb123) [reply include/cuco/bloom_filter_policies.cuh]: Added notes!

**Review by PointKernel — APPROVED** [review]

**Review by sleeepyjack — APPROVED** [review]

**Pull request merged**